pkg/app/carbonapi/http_handlers.go

```diff
@@ -939,39 +940,148 @@ func (app *App) findHandler(w http.ResponseWriter, r *http.Request, lg *zap.Logg
 		logLevel = zapcore.ErrorLevel
 		return
 	}
-
+	// TODO: extend writeResponse w/new formats and migrate to it
 	if contentType == jsonFormat && jsonp != "" {
-		w.Header().Set("Content-Type", contentTypeJavaScript)
-		if _, writeErr := w.Write([]byte(jsonp)); writeErr != nil {
-			toLog.HttpCode = 499
-			logLevel = zapcore.WarnLevel
-			return
-		}
-		if _, writeErr := w.Write([]byte{'('}); writeErr != nil {
+		if writeResponse(ctx, w, blob, jsonFormat, jsonp) != nil {
 			toLog.HttpCode = 499
 			logLevel = zapcore.WarnLevel
 			return
 		}
+	} else {
+		w.Header().Set("Content-Type", contentType)
 		if _, writeErr := w.Write(blob); writeErr != nil {
 			toLog.HttpCode = 499
 			logLevel = zapcore.WarnLevel
 			return
 		}
-		if _, writeErr := w.Write([]byte{')'}); writeErr != nil {
-			toLog.HttpCode = 499
-			logLevel = zapcore.WarnLevel
-			return
+	}
+
+	toLog.HttpCode = http.StatusOK
+}
+
+func expandEncoder(globs dataTypes.Matches, leavesOnly bool, groupByExpr bool) ([]byte, error) {
+	var b bytes.Buffer
+	groups := make(map[string][]string)
+	seen := make(map[string]struct{})
+	nodeCount := len(strings.Split(globs.Name, "."))
+	names := make([]string, 0, len(globs.Matches))
+	for _, g := range globs.Matches {
+		if leavesOnly && !g.IsLeaf {
+			continue
+		}
+		name := g.Path
+		nodes := strings.SplitN(name, ".", nodeCount+1)
+		if len(nodes) > nodeCount {
+			name = strings.Join(nodes[:nodeCount], ".")
+		} else {
+			name = strings.Join(nodes, ".")
+		}
+		if _, ok := seen[name]; ok {
+			continue
+		}
+		seen[name] = struct{}{}
+		names = append(names, name)
+	}
+	sort.Strings(names)
+	groups[globs.Name] = names
+	data := map[string]interface{}{
+		"results": groups,
+	}
+	if !groupByExpr {
+		flatData := make([]string, 0)
+		for _, group := range groups {
+			flatData = append(flatData, group...)
 		}
+		data["results"] = flatData
+	}
+	err := json.NewEncoder(&b).Encode(data)
+	return b.Bytes(), err
+}
+
+func (app *App) expandHandler(w http.ResponseWriter, r *http.Request, lg *zap.Logger) {
+	t0 := time.Now()
+	defer func() {
+		d := time.Since(t0).Seconds()
+		app.ms.DurationTotal.WithLabelValues("expand").Observe(d)
+		app.ms.ExpandDurationExp.Observe(d)
+		app.ms.ExpandDurationLin.Observe(d)
+	}()
+
+	ctx, cancel := context.WithTimeout(r.Context(), app.config.Timeouts.Global)
+	defer cancel()
+	uuid := util.GetUUID(ctx)
+
+	app.ms.Requests.Inc()
+
+	query := r.FormValue("query")
+	leavesOnly := !parser.TruthyBool(r.FormValue("leavesOnly"))
+	groupByExpr := !parser.TruthyBool(r.FormValue("groupByExpr"))
```

Comment:
I guess we should not have ! (not) on this line, as well.

Reply:
yes, good catch again, thanks, Emad!